src/components/TextInput/TextInput.js

```diff
@@ -1,104 +1,190 @@
 import PropTypes from 'prop-types';
 import React from 'react';
 import classNames from 'classnames';
+import Icon from '../Icon';
 
-const TextInput = ({
-  labelText,
-  className,
-  id,
-  placeholder,
-  type,
-  onChange,
-  onClick,
-  hideLabel,
-  invalid,
-  invalidText,
-  helperText,
-  light,
-  ...other
-}) => {
-  const textInputProps = {
-    id,
-    onChange: evt => {
-      if (!other.disabled) {
-        onChange(evt);
-      }
-    },
-    onClick: evt => {
-      if (!other.disabled) {
-        onClick(evt);
-      }
-    },
-    placeholder,
-    type,
+export default class TextInput extends React.Component {
+  state = {
+    type: this.props.type,
```

Comment:
Changing this component to stateful means that you need to take care of prop -> state sync (coping with change in `type` prop). Two ways to address them:

1. [`getDerivedStateFromProps()`](https://reactjs.org/docs/react-component.html#static-getderivedstatefromprops) (search for the method in our codebase to see how)
2. Keep this component stateless and let consumer to handle show/hide button's event to change `type` prop

Reply:
yeah I am still unsure about how I want to handle state with this component. Ideally I would like to keep this component stateless but I don't know if the consumer should be handling the `type` prop change. I think I would prefer to keep the `type` change local to `<TextInput>`, so I will go down the `getDerivedStateFromProps` route. What do you think @asudoh ?